blueprint/src/code/tu.tex

```diff
@@ -3,49 +3,104 @@ \section{TU Matrices}
 \end{lemma}
 
 \begin{proof}[Proof sketch]
   \uses{def:code_tu_matrix}
-  Gaussian elimination. Basis submatrix: its columns form a basis of all columns, its rows form a basis of all rows.
+  \leanok
+  Any square submatrix of $B$ is a submatrix of $A$, so its determinant is $0$ or $\pm 1$. Thus, $B$ is TU.
 \end{proof}
 
-\begin{lemma}[any submatrix of a TU matrix is TU]
-  \label{lem:code_submatrix_of_tu}
+\begin{lemma}[transpose of TU is TU]
+  \label{lem:code_tu_transpose}
   \uses{def:code_tu_matrix}
-  \lean{Matrix.TU.submatrix}
-  Let $A$ be a real matrix that is TU and let $B$ be a submatrix of $A$. Then $B$ is TU.
+  \lean{Matrix.TU.transpose}
+  \leanok
+  Let $A$ be a TU matrix. Then $A^{T}$ is TU.
 \end{lemma}
 
 \begin{proof}[Proof sketch]
   \uses{def:code_tu_matrix}
-  Any square submatrix of $B$ is a submatrix of $A$, so its determinant is $0$ or $\pm 1$. Thus, $B$ is TU.
+  \leanok
+  A submatrix $T$ of $A^{T}$ is a transpose of a submatrix of $A$, so $\det T \in \{0, \pm 1\}$.
+\end{proof}
+
+\begin{lemma}[appending zero vector to TU]
+  \label{lem:code_tu_add_zero_row}
+  \uses{def:code_tu_matrix}
+  \lean{Matrix.TU_adjoin_row0s_iff}
+  \leanok
+  Let $A$ be a matrix. Let $a$ be a zero row. Then $C = \left[ A / a \right]$ is TU exactly when $A$ is.
+\end{lemma}
+
+\begin{proof}[Proof sketch]
+  \uses{def:code_tu_matrix,lem:code_submatrix_of_tu}
+  \leanok
+  Let $T$ be a square submatrix of $C$, and suppose $A$ is TU. If $T$ contains a zero row, then $\det T = 0$. Otherwise $T$ is a submatrix of $A$, so $\det T \in \{0, \pm 1\}$.
+  For the other direction, because $A$ is a submatrix of $C$, we can apply lemma \ref{lem:code_submatrix_of_tu}.
+\end{proof}
+
+\begin{lemma}[appending unit vector to TU]
+  \label{lem:code_tu_add_unit_row}
+  \uses{def:code_tu_matrix}
+  \lean{Matrix.TU_adjoin_rowUnit}
+  \leanok
+  Let $A$ be a matrix. Let $a$ be a unit row. Then $C = \left[ A / a \right]$ is TU exactly when $A$ is.
+\end{lemma}
+
+\begin{proof}[Proof sketch]
+  \uses{def:code_tu_matrix,lem:code_submatrix_of_tu}
+  Let $T$ be a square submatrix of $C$, and suppose $A$ is TU.
+  If $T$ contains the $\pm 1$ entry of the unit row,
+  then $\det T$ equals the determinant of some submatrix of $A$ times $\pm 1$,
+  so $\det T \in \{0, \pm 1\}$.
+  If $T$ contains some entries of the unit row except the $\pm 1$, then $\det T = 0$.
+  Otherwise $T$ is a submatrix of $A$, so $\det T \in \{0, \pm 1\}$.
+  For the other direction, simply note that $A$ is a submatrix of $C$, and use lemma \ref{lem:code_submatrix_of_tu}.
+\end{proof}
+
+\begin{lemma}[TUness with adjoint identity matrix]
+  \label{lem:code_tu_adjoin_id}
+  \uses{def:code_tu_matrix}
+  \lean{Matrix.TU_adjoin_id_below_iff,Matrix.TU_adjoin_id_above_iff,Matrix.TU_adjoin_id_left_iff,Matrix.TU_adjoin_id_right_iff}
+  \leanok
+  $A$ is TU iff every basis matrix of $\left[ I \mid A \right]$ has determinant $\pm 1$.
+\end{lemma}
```

Comment:
technically this lemma doesn't quite agree with the lean statement. if you're ok with it, we can change the blueprint to more accurately reflect what is proven in lean.

Reply:
I'll leave it up to you.